### PR TITLE
Use flashinfer kernel for autoregressive sampling

### PR DESCRIPTION
**Pull request by kasohrab**
Use flashinfer kernel for autoregressive sampling instead of torch. Verified output still looks ok from offline_inference.py.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to Vidur! Before submitting the pull request, please ensure the PR meets the following criteria. This helps Vidur maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Profiling]</code> For changes on the profiling module. </li>
    <li><code>[Core]</code> for changes in the core simulator logic </li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <code>make format</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of Vidur. It helps user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to Vidur. Your contributions make Vidur a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by kasohrab** [comment]
> @kasohrab please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
>  > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree

File: sarathi/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -219,3 +219,18 @@ def _sample(
             outputs[seq_idx] = SamplerOutput(seq_id, sample_result)
 
     return outputs
+
+
+def _top_k_top_p_with_flashinfer(
+    logits: torch.Tensor, top_ks: torch.Tensor, top_ps: torch.Tensor
+):
+    max_top_k_round = 32
```
Comment (AgrawalAmey) [inline sarathi/model_executor/layers/sampler.py:227]: why 32?

File: sarathi/model_executor/layers/sampler.py [hunk 2]
```diff
@@ -59,21 +64,16 @@ def forward(
...
-        # Use log_softmax to ensure numerical stability.
-        logprobs = torch.log_softmax(logits, dim=-1, dtype=torch.float)
+        flashinfer_sample_result = _top_k_top_p_with_flashinfer(logits, top_ks, top_ps)
 
-        # Sample the next tokens.
-        return _sample(probs, logprobs, seq_metadata_list)
+        outputs = []
+        for i, seq_metadata in enumerate(seq_metadata_list):
+            seq_id = seq_metadata.seq.seq_id
+            outputs.append(SamplerOutput(seq_id, flashinfer_sample_result[i]))
```
Comment (AgrawalAmey) [inline sarathi/model_executor/layers/sampler.py:75]: nit: please do this in list comprehension for efficiency

File: sarathi/model_executor/layers/sampler.py [hunk 3]
```diff
@@ -59,21 +64,16 @@ def forward(
...
+        top_ps = torch.tensor(top_ps, dtype=logits.dtype, device=logits.device)
+        top_ks = torch.tensor(top_ks, dtype=torch.int, device=logits.device)
 
-        # We use float32 for probabilities and log probabilities.
-        # Compute the probabilities.
-        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
-        # Compute the log probabilities.
-        # Use log_softmax to ensure numerical stability.
-        logprobs = torch.log_softmax(logits, dim=-1, dtype=torch.float)
+        flashinfer_sample_result = _top_k_top_p_with_flashinfer(logits, top_ks, top_ps)
```
Comment (AgrawalAmey) [inline sarathi/model_executor/layers/sampler.py:70]: do a .cpu on the output -- this should fix the perf regression

File: sarathi/model_executor/layers/sampler.py (hunk 1 above)
Comment (kasohrab) [reply sarathi/model_executor/layers/sampler.py:226]: It's the same value used in the vllm codebase. No explanation was provided there either so not sure

Comment (AgrawalAmey) [reply sarathi/model_executor/layers/sampler.py:226]: i see, i guess they really don't want it to fail, in any case, can you please move this constant to the top of the file in uppercase

File: sarathi/model_executor/layers/sampler.py [hunk 4]
```diff
@@ -59,21 +62,17 @@ def forward(
 
         # Apply top-p and top-k truncation.
         top_ps, top_ks = _get_top_p_top_k(seq_metadata_list, self.vocab_size)
-        assert len(top_ps) == len(top_ks) == logits.shape[0]
-        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
-        do_top_k = any(k != self.vocab_size for k in top_ks)
-        if do_top_p or do_top_k:
```
Comment (AgrawalAmey) [inline sarathi/model_executor/layers/sampler.py:65]: this looks like an important optimization, can we use non top-k/top-p kernel when neither top-p/top-k is requried?

File: sarathi/model_executor/layers/sampler.py (hunk 1 above)
Comment (kasohrab) [reply sarathi/model_executor/layers/sampler.py:226]: Ok will do that. Also realized 32 is reccomend by flashinfer documentation: https://docs.flashinfer.ai/generated/flashinfer.sampling.top_k_top_p_sampling_from_logits.html

File: sarathi/model_executor/layers/sampler.py (hunk 4 above)
Comment (kasohrab) [reply sarathi/model_executor/layers/sampler.py:65]: Looking further I am still inclined to say it wouldn't optimize beyond what flash infer does: https://github.com/flashinfer-ai/flashinfer/blob/46ec89317fb112dfb04bc740e711ccd3cdcf96f8/include/flashinfer/sampling.cuh#L566. But let me know if I am mistaken (that is the kernel ultimately called by the logits method too).

Comment (kasohrab) [reply sarathi/model_executor/layers/sampler.py:65]: But I can investigate if this code path plus https://docs.flashinfer.ai/generated/flashinfer.sampling.sampling_from_probs.html would be more efficient.

Comment (AgrawalAmey) [reply sarathi/model_executor/layers/sampler.py:65]: i see, whatever seems more reasonable -- might be easier to benchmark and compare than analyzing code potentially. The PR looks pretty good to me overall, can you please share the perf numbers from benchmark.main, if everything is looking good, let's quickly merge this. cc @nitinkedia7

File: sarathi/model_executor/layers/sampler.py (hunk 1 above)
Comment (yzh119) [reply sarathi/model_executor/layers/sampler.py:226]: Hi @AgrawalAmey @kasohrab , thanks for integrating this, I hope the sampling kernel would benefit your project!

The reason it's set to 32 is that we implement a rejection sampling algorithm where each round expect to filter out around half of the choices, and hopefully 32 rounds could find the 1/(2^32) choice (I hope the vocabulary size would not exceed this value). In most cases, the algorithm stops much earlier. But 32 is a safe choice.

Comment (AgrawalAmey) [reply sarathi/model_executor/layers/sampler.py:226]: @yzh119 thanks a lot for sharing the the explanation. The project has greatly benefited from the availability of  flashinfer, thanks again!

**Comment by AgrawalAmey** [comment]
@kasohrab are we good to merge this PR? cc @nitinkedia7

**Comment by kasohrab** [comment]
> @kasohrab are we good to merge this PR? cc @nitinkedia7

Tested bs=512, sarathi scheduler and seeing ~10% improvement in sampling e2e with flashinfer. So looks good from my end!
Screenshot 2024-09-10 at 13 04 04

**Comment by AgrawalAmey** [comment]
@kasohrab can you please remove dead codepaths?

**Comment by AgrawalAmey** [comment]
@nitinkedia7 can we please merge this request

**Review by nitinkedia7 — APPROVED** [review]
